Survival-DEV/blog-backend
# Blog-33: send confirmation email

### :dart:  What is been done in this PR:

- [x] Add a register method, that sends a user email after successful signup.
- [x] Create a dynamic template using `sendGrid` for email servicing. 
- [x]  Redirect the user to confirm the user after clicking on the email verification button.
- [x] Add methods for confirming the account & resend another verification email
- [x] Update the userEntity with a new column `isEmailConfirmed`
- [x] clean up the code
 **Related to #33**

### in progress
- [x] Debug the current issue of failed `jwtService.verify` service, here:
https://github.com/Survival-DEV/blog-backend/blob/96fd3b65f1c5056d2692402c035ae47aa11677f5/src/modules/notifications/notifications.service.ts#L57-L61
the problem was with `configService`, importing it directly from `.env` solved the issue
-----
## UI Elements:
![image](https://user-images.githubusercontent.com/56412800/130684266-e888d511-13aa-468a-a49d-8df0a0004be0.png)

----
### :compass:   How to test it
Steps to test my PR:
1. `npm run ci`
2. make sure to setup your `env` file to have these params:
    <details>
          <summary>Expand to see the params</summary>
        
          > DATABASE_URL=<your db url#>
          > NODE_ENV=DEBUG  //for now
          > logDB=["query", "error"]
          > PORT=<your port#>
          > TRUSTED_IP=["127.0.0.1"]
          > WHITELIST=['http://localhost']
          > ALLOWED_METHODS='GET,HEAD,PUT,PATCH,POST,DELETE,OPTIONS'
          > JWT_VERIFICATION_TOKEN_SECRET='secreto2'
          > JWT_VERIFICATION_TOKEN_EXPIRATION_TIME='60s'
          > EMAIL_CONFIRMATION_URL='http://localhost:5000/api/confirm-email/'
          > SEND_GRID_ACCESS_KEY=`<your sendgird key>`
          > FROM_EMAIL=`your sendgrid email`
          > SENDGRID_TEMPLATE_ID=d-3285e7571ee64e8fafedb163996e5a55
       </details>

4. `docker-compose up`   --to run redis 
3. `npm run start:dev`   --in another terminal
